test/unit/org/apache/cassandra/io/sstable/format/SSTableFlushObserverTest.java

```diff
@@ -42,189 +48,507 @@
+        public void staticRow(Row staticRow, long position)
         {
-            currentKey = Pair.create(key.getKey(), indexPosition);
+            currentStaticRow = staticRow;
+            currentStaticRowPosition = position;
         }
 
         @Override
-        public void nextUnfilteredCluster(Unfiltered row)
+        public void nextUnfilteredCluster(Unfiltered unfiltered, long position)
         {
-            if (row.isRow())
-                ((Row) row).forEach((c) -> rows.put(currentKey, (Cell<?>) c));
+            unfiltereds.add(new UnfilteredEntry(unfiltered, position));
         }
 
         @Override
         public void complete()
         {
             isComplete = true;
+            if (pendingHeader)
+                headers.add(new HeaderEntry());
+        }
+
+        class HeaderEntry
+        {
+            final DecoratedKey key;
+            final long keyPosition;
+            final DeletionTime deletionTime;
+            final long deletionTimePosition;
+            final Row staticRow;
+            final long staticRowPosition;
+
+            HeaderEntry()
+            {
+                this.key = currentKey;
+                this.keyPosition = currentKeyPosition;
+                this.deletionTime = currentDeletionTime;
+                this.deletionTimePosition = currentDeletionTimePosition;
+                this.staticRow = currentStaticRow;
+                this.staticRowPosition = currentStaticRowPosition;
+            }
+        }
+
+        class UnfilteredEntry
+        {
+            final DecoratedKey key;
+            final long keyPosition;
+            final long unfilteredPosition;
+            final Unfiltered unfiltered;
+
+            UnfilteredEntry(Unfiltered unfiltered, long unfilteredPosition)
+            {
+                assertTrue(!unfiltered.isRow() || !((Row) unfiltered).isStatic());
+                this.key = currentKey;
+                this.keyPosition = currentKeyPosition;
+                this.unfilteredPosition = unfilteredPosition;
+                this.unfiltered = unfiltered;
+            }
+        }
+    }
+
+    private static class PartitionHeader
+    {
+        final DecoratedKey key;
+        final DeletionTime deletionTime;
+        final Row staticRow;
+
+        PartitionHeader(DecoratedKey key, DeletionTime deletionTime, Row staticRow)
+        {
+            this.key = key;
+            this.deletionTime = deletionTime;
+            this.staticRow = staticRow;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            PartitionHeader that = (PartitionHeader) o;
+            return Objects.equals(key, that.key) &&
+                   Objects.equals(deletionTime, that.deletionTime) &&
+                   Objects.equals(staticRow, that.staticRow);
```

Comment:
nit: Could call the `equals()` methods directly here if you wanted to, ex. `staticRow.equals(that.staticRow)`